File: 2563. Count the Number of Fair Pairs/Solution.cpp
#include <algorithm>
#include <ranges>
#include <vector>
using namespace std;

class Solution {
public:
    long long countFairPairs(vector<int>& nums, int lower, int upper)
    {
        ranges::sort(nums);
        long long ans = 0;
        int left = nums.size(), right = nums.size(), n = nums.size();
        for (int j = 0; j < n; ++j) {
            while (left && nums[left - 1] >= lower - nums[j])
                --left;
            while (right && nums[right - 1] > upper - nums[j])
                --right;
            ans += min(right, j) - min(left, j);
        }
        return ans;
    }
};

File: 2563. Count the Number of Fair Pairs/Solution.py
from typing import List


class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        ans = 0
        nums.sort()
        left = right = len(nums)
        for j, x in enumerate(nums):
            while left and nums[left - 1] >= lower - x:
                left -= 1
            while right and nums[right - 1] > upper - x:
                right -= 1
            ans += min(right, j) - min(left, j)
        return ans
